### DISCUSSION TITLE
so-firewall cmd line incorrect for --apply

### DISCUSSION THREAD

**Discussion by trb143** [Security-Onion-Solutions/securityonion#11508] [2.4]
### Version

2.4.20

### Installation Method

Security Onion ISO image

### Description

other (please provide detail below)

### Installation Type

Standalone

### Location

on-prem with Internet access

### Hardware Specs

Meets minimum requirements

### CPU

16

### RAM

32

### Storage for /

500gb

### Storage for /nsm

1000gb

### Network Traffic Collection

span port

### Network Traffic Speeds

1Gbps to 10Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

Yes, there are salt failures (please provide detail below)

### Logs

Yes, there are additional clues in /opt/so/log/ (please provide detail below)

### Detail

Having issues accessing the SOC from a different subnet I accessed the box directly.
Running the so-firewall command a number of times to get the subnets added two issues stood out.

Should I add to analyst or standalone. - Added to both - an update to documentation may be needed here.

so-filewall --apply         does not work.
There is a check for 0 parameters which kicks you out the script.
so-firewall --apply apply            does work.

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by dougburks** [comment]
> Should I add to analyst or standalone. - Added to both - an update to documentation may be needed here.

Have you tried it as shown at https://docs.securityonion.net/en/2.4/firewall.html#configuration?
_If for some reason you can’t access [Security Onion Console (SOC)](https://docs.securityonion.net/en/2.4/soc.html#soc), you can use the so-firewall command to allow your IP address to connect (replacing IPADDRESS with your actual IP address):_
```
so-firewall includehost analyst IPADDRESS
```

> so-filewall --apply does not work.
There is a check for 0 parameters which kicks you out the script.
so-firewall --apply apply does work.

Here's the usage:
```
so-firewall
Usage: /sbin/so-firewall [OPTIONS] <COMMAND> [ARGS...]
  Options:
   --apply         - After updating the firewall configuration files, apply the new firewall state

  General commands:
    help           - Prints this usage information.
    apply          - Apply the firewall state.

  Host commands:
    includehost    - Includes the given IP in the given group. Args: <GROUP_NAME> <IP>
    addhostgroup   - Adds a new, custom host group. Args: <GROUP_NAME>

  Where:
   GROUP_NAME    - The name of an alias group (Ex: analyst)
   IP            - Either a single IP address (Ex: 8.8.8.8) or a CIDR block (Ex: 10.23.0.0/16).
```

So I think probably what you want to do is combine `--apply` with your `includehost` like this:
```
so-firewall --apply includehost analyst IPADDRESS
```

**Comment by trb143** [comment]
Looking at the code --apply and includehost are different commands.  It runs one or the other not both.

--apply does not need any arguments to run as all the functions are None. 

Looking at the code again the advertised option addhostgroup will also fail as the array:
  commands = {
    "help": showUsage,
    "includehost": includehost,
    "apply": apply
  }

is missing that option.

**Reply by jertel** [reply]
They are not the same. 

`--apply` is an option that can accompany a command.

whereas

`apply` (no leading hyphens) is a command.

Using the `apply` (without hyphens) as a command tells `so-firewall` you are only wanting to apply the firewall state, but not make any IP additions or removals. Using the `--apply` as an argument tells `so-firewall` that when the requested command is complete immediately apply the firewall state before exiting.
